File: common/src/main/java/com/yuji/xlt/common/api/RetrofitManager.java
package com.yuji.xlt.common.api;

import com.yuji.xlt.ability.utils.Logger;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * 网络请求封装<BR>
 *
 * @author dengshishuang
 * @version [V1.0.0, 2021/7/28]
 * @since V1.0.0
 */
public class RetrofitManager {
    private static final String BASEURL = "http://10.10.128.7:18888/";
    private static final int TIMEOUT = 5000;
    private static Retrofit retrofit;
    private static volatile RetrofitManager retrofitManager;

    private RetrofitManager() {
        retrofit = getRetrofit();
    }

    public static RetrofitManager getInstance() {
        if (retrofitManager == null) {
            synchronized (RetrofitManager.class) {
                if (retrofitManager == null) {
                    retrofitManager = new RetrofitManager();
                }
            }
        }
        return retrofitManager;
    }

    public <T> T create(Class<T> t) {
        return retrofit.create(t);
    }

    private Retrofit getRetrofit() {
        return new Retrofit.Builder()
                .baseUrl(BASEURL)
                .client(getOkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build();
    }

    private OkHttpClient getOkHttpClient() {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
            @Override
            public void log(String message) {
                Logger.d(message);
            }
        });
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        return new OkHttpClient.Builder()
                .connectTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
                .addInterceptor(loggingInterceptor)
                .build();
    }
}

File: common/src/main/java/com/yuji/xlt/common/http/RetrofitManager.kt
package com.yuji.xlt.common.http

/**
 * 初始化Retrofit,维护多个ApiService单例对象<BR>
 *
 * @author dengshishuang
 * @version [V1.0.0, 2021/8/30]
 * @since V1.0.0
 */
open class RetrofitManager {
    companion object {
        /**
         * 用于存储ApiService
         */
        private val map = mutableMapOf<Class<*>, Any>()

        /**
         * 只初始化一次
         */
        private val retrofit = RetrofitFactory.factory()

        //动态指定域名
        fun <T : Any> getApiService(apiClass: Class<T>): T {
            return getService(apiClass)
        }

        //动态指定域名
        fun <T : Any> getResApiService(apiClass: Class<T>): T {
            return getService(apiClass)
        }

        /**
         * 获取ApiService单例对象
         */
        private fun <T : Any> getService(apiClass: Class<T>): T {
            //重入锁单例避免多线程安全问题
            return if (map[apiClass] == null) {
                synchronized(RetrofitManager::class.java) {
                    val t = retrofit.create(apiClass)
                    if (map[apiClass] == null) {
                        map[apiClass] = t
                    }
                    t
                }
            } else {
                map[apiClass] as T
            }
        }
    }
}